src/common/snippets/src/pass/lowered/insert_loops.cpp

```diff
@@ -0,0 +1,132 @@
+    const auto& loop_manager = linear_ir.get_loop_manager();
+    auto loop_depth = lowering_config.m_loop_depth;
+
+    // Parameters Results or Constants are ignored. They can't be used as a loop starting point
+    auto is_not_start_point = [](const std::shared_ptr<ov::Node>& node) {
+        return ov::is_type<opset1::Result>(node) ||
+               ov::is_type<opset1::Constant>(node) ||
+               ov::is_type<opset1::Parameter>(node) ||
+               ov::is_type<opset1::Softmax>(node) ||
+               ov::is_type<op::Brgemm>(node);
+    };
+
+    // Scalars cannot be a single node in Loop
+    // So after common markup we should manually mark Scalars and
+    // explicitly move to before the corresponding consumer
+    std::vector<LoweredExprIR::constExprIt> scalars_iterators;
+
+    for (auto expr_it = linear_ir.cbegin(); expr_it != linear_ir.cend(); expr_it++) {
+        const auto expr = *expr_it;
+        const auto& node = expr->get_node();
+        if (ov::is_type<opset1::Constant>(node)) {
+            scalars_iterators.push_back(expr_it);
+            continue;
+        }
+        if (is_not_start_point(node))
+            continue;
+
+        auto loop_begin_pos = expr_it;
+        auto loop_end_pos = loop_begin_pos;
+
+        const auto& outputs = expr->get_outputs();
+        const auto& loop_inner_layout = outputs.front()->get_layout();
+        const auto& loop_inner_subtensor = outputs.front()->get_subtensor();
+
+        std::vector<LoweredExprPtr> body_exprs;
+        bool is_inside = true;
+        do {
+            const auto& prev_node = loop_end_pos->get()->get_node();
+            body_exprs.push_back(*loop_end_pos);
+            loop_end_pos++;
+            // If iterator is the last, we should finish Loop
+            if (loop_end_pos == linear_ir.end())
+                break;
+
+            // If iterator is the last, we should finish Loop
+            const auto& current_expr = *loop_end_pos;
+            const auto& current_node = current_expr->get_node();
+            if (ov::is_type<op::Brgemm>(current_node) ||
+                ov::is_type<opset1::Softmax>(current_node) ||
+                ov::is_type<opset1::Result>(current_node) ||
+                ov::is_type<opset1::Constant>(current_node))
+                break;
+
+            // If the next expr isn't real customer of prev expr we should finish Loop
+            std::set<LoweredExprPtr> prev_node_customers;
+            const auto prev_outputs = body_exprs.back()->get_outputs();
+            for (const auto& td_output : prev_outputs) {
+                const auto consumers = linear_ir.get_exprs_by_input(td_output);
+                prev_node_customers.insert(consumers.begin(), consumers.end());
+            }
+            auto compare = [&current_node](const LoweredExprPtr& consumer){ return consumer->get_node() == current_node;};
+            if (std::none_of(prev_node_customers.begin(), prev_node_customers.end(), compare))
```

Comment:
You don't have to create a custom set here. It's probably easier to perform the connectivity check based on the current exprs inputs. Please, take a look:
```suggestion
            auto connected = [&](const TensorDescriptorPtr& td) {return linear_ir.get_expr_by_output(td) == body_exprs.back();};
            if (std::none_of(ins.begin(), ins.end(), connected))
                break;
```

Reply:
Good catch! Thanks!